ttl/ttlworker/job_manager.go

```diff
@@ -0,0 +1,582 @@
+			newJobTables = newJobTables[1:]
+			logutil.Logger(m.ctx).Debug("try lock new job", zap.Int64("tableID", table.ID))
+			job, err = m.lockNewJob(m.ctx, se, table, now)
+			if job != nil {
+				m.appendJob(job)
+			}
+		}
+		if err != nil {
+			logutil.Logger(m.ctx).Warn("fail to create new job", zap.Error(err))
+		}
+		if job == nil {
+			continue
+		}
+
+		for !job.AllSpawned() {
+			task, err := job.peekScanTask()
+			if err != nil {
+				logutil.Logger(m.ctx).Warn("fail to generate scan task", zap.Error(err))
+				break
+			}
+
+			for len(idleScanWorkers) > 0 {
+				idleWorker := idleScanWorkers[0]
+				idleScanWorkers = idleScanWorkers[1:]
+
+				err := idleWorker.Schedule(task)
+				if err != nil {
+					logutil.Logger(m.ctx).Info("fail to schedule task", zap.Error(err))
+					continue
+				}
+
+				ctx, cancel := context.WithTimeout(m.ctx, ttlInternalSQLTimeout)
+				err = job.changeStatus(ctx, se, cache.JobStatusRunning)
+				if err != nil {
+					// not a big problem, current logic doesn't depend on the job status to promote
+					// the routine, so we could just print a log here
+					logutil.Logger(m.ctx).Error("change ttl job status", zap.Error(err), zap.String("id", job.id))
+				}
+				cancel()
+
+				logArgs := []zap.Field{zap.String("table", task.tbl.TableInfo.Name.L)}
+				if task.tbl.PartitionDef != nil {
+					logArgs = append(logArgs, zap.String("partition", task.tbl.PartitionDef.Name.L))
+				}
+				logutil.Logger(m.ctx).Debug("schedule ttl task",
+					logArgs...)
+
+				job.nextScanTask()
+				break
+			}
+
+			if len(idleScanWorkers) == 0 {
+				break
+			}
+		}
+	}
+}
+
+func (m *JobManager) idleScanWorkers() []scanWorker {
+	workers := make([]scanWorker, 0, len(m.scanWorkers))
+	for _, w := range m.scanWorkers {
+		if w.(scanWorker).Idle() {
+			workers = append(workers, w.(scanWorker))
+		}
+	}
+	return workers
+}
+
+func (m *JobManager) localJobs() []*ttlJob {
+	for _, job := range m.runningJobs {
+		status := m.tableStatusCache.Tables[job.tbl.ID]
+		if status == nil || status.CurrentJobOwnerID != m.id {
+			m.removeJob(job)
+			continue
+		}
+	}
+	return m.runningJobs
+}
+
+// readyForNewJobTables returns all tables which should spawn a TTL job according to cache
+func (m *JobManager) readyForNewJobTables(now time.Time) []*cache.PhysicalTable {
+	tables := make([]*cache.PhysicalTable, 0, len(m.infoSchemaCache.Tables))
+	for _, table := range m.infoSchemaCache.Tables {
+		status := m.tableStatusCache.Tables[table.ID]
+		ok := m.couldTrySchedule(status, now)
+		if ok {
+			tables = append(tables, table)
+		}
+	}
+
+	return tables
+}
+
+// couldTrySchedule returns whether a table should be tried to run TTL
+func (m *JobManager) couldTrySchedule(table *cache.TableStatus, now time.Time) bool {
+	if table == nil {
+		// if the table status hasn't been created, return true
+		return true
+	}
+	if table.CurrentJobOwnerID != "" {
+		// see whether it's heart beat time is expired
+		hbTime := table.CurrentJobOwnerHBTime
```

Comment:
seems when a job's hbtime stops to update, a new job with a new ttl expire will be created. I'm not sure whether we should make a little more complex strategy to make the job can be "resume" instead a new job. This maybe useful in some cases, for example, if we have a bug that a job will crash after running for 1hour, it still can guarantee a job will be finished finally and make sure the "cleared watermark"  can be pushed forward.

Reply:
If we want to implement the "resume", another thing we should concern is that if a very old job failed, we should not resume it.